### PR TITLE
S1125: Faulty code fix on C# pattern matching in conditional operator

### LINKED ISSUE #2618: S1125: Faulty code fix on C# pattern matching in conditional operator
### Description

A bit of an exotic error. The following conditional expression

`var isSuccess = y is ISuccessEvaluator evaluator ? evaluator.Evaluate(result) : true;`

results in a S1125 warning "remove the unnecessary boolean literal(s)" . The suggested code fix would result in:

`var isSuccess = !y is ISuccessEvaluator evaluator || evaluator.Evaluate(result);`

which produces a compiler error "! cannot be applied to operand of type "[whatever type y is]".

### Expected behavior

`var isSuccess = !(y is ISuccessEvaluator evaluator) || evaluator.Evaluate(result);`

is the code fix that will not throw a compiler error. The braces are necessary in this instance because otherwise the compiler will try to apply the negation to type `y` which is not a bool.  

### Known workarounds

Add the braces manually after applying the code fix

### Related information

* SonarC# Version 7.16.0.8981
* Visual Studio Version VS 2019 Version 16.2.0 Preview 3.0

### LINKED ISSUE #7462: Fix S1125: Codefix should set correct condition priority by adding parentheses.
The statement
```csharp
obj == null ? false : obj.ToString() == "x" || obj.ToString() == "y";
```
get fixed to:
```csharp
!(obj == null) && obj.ToString() == "x" || obj.ToString() == "y";
```

which throws null exception because the condition is evaluated as `(condition1 && conction2) || condition3`. `(condition1 && conction2)` is `false` and so `condition3` will be evaluated.

It should be fixed to:
```csharp
!(obj == null) && (obj.ToString() == "x" || obj.ToString() == "y");
```

so that the evaluation is `condition1 && (condition2 || condition3)` and if `condition1` is false then `(condition2 || condition3)` won't be evaluated at all.

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Fixes #2618
Fixes #7462

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs [hunk 1]
```diff
@@ -449,6 +449,30 @@ public static ConditionalAccessExpressionSyntax GetRootConditionalAccessExpressi
...
+            {
+            { RawKind: (int)SyntaxKind.FalseLiteralExpression } => true, // false
+            { RawKind: (int)SyntaxKind.LogicalNotExpression } => IsTrue(((PrefixUnaryExpressionSyntax)node).Operand), // !true
+            { RawKind: (int)SyntaxKindEx.ConstantPattern } => IsFalse(((ConstantPatternSyntaxWrapper)node).Expression), // is false
+            { RawKind: (int)SyntaxKindEx.NotPattern } => IsTrue(((UnaryPatternSyntaxWrapper)node).Pattern), // is not true
+            { RawKind: (int)SyntaxKind.ParenthesizedExpression } => IsFalse(((ParenthesizedExpressionSyntax)node).Expression), // (false)
+            { RawKind: (int)SyntaxKindEx.ParenthesizedPattern } => IsFalse(((ParenthesizedPatternSyntaxWrapper)node).Pattern), // is (false)
+            _ => false,
+        };
+
```
Comment (cristian-ambrosini-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Extensions/SyntaxNodeExtensions.cs:475]: Methods from #7687 (not merged yet)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/BooleanLiteralUnnecessaryCodeFix.cs [hunk 2]
```diff
@@ -250,73 +250,66 @@ private static Document RemoveConditional(Document document, SyntaxNode root, Co
...
-               || expression is ConditionalExpressionSyntax)
-            {
-                exp = SyntaxFactory.ParenthesizedExpression(expression);
-            }
+        private static ExpressionSyntax GetNegatedExpression(ExpressionSyntax expression) =>
+            SyntaxFactory.PrefixUnaryExpression(SyntaxKind.LogicalNotExpression, ParenthesizeIfComplex(expression));
 
-            return SyntaxFactory.PrefixUnaryExpression(SyntaxKind.LogicalNotExpression, exp);
-        }
+        private static ExpressionSyntax ParenthesizeIfComplex(ExpressionSyntax expression) =>
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/BooleanLiteralUnnecessaryCodeFix.cs:309]: Easier: Alyways Parenthsize and add the `Simplifier.Annotation` to the newly created expression. Roslyn should remove the parenthesis if not needed.
See also https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.simplification.simplifier?view=roslyn-dotnet-4.6.0

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Give the simplifier annotation a try.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/BooleanLiteralUnnecessaryCodeFix.cs (hunk 2 above)
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/BooleanLiteralUnnecessaryCodeFix.cs:309]: From what I understood, you need to pass to `ReduceAsync` method the full document, marking with `Annotation` the sub-tree which you are interested to reduce. Then I wonder how to retrieve back the same reduced SyntaxNode from the full document, any ideas?

Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/BooleanLiteralUnnecessaryCodeFix.cs:309]: Sorry, with `.WithAdditionalAnnotations(Simplifier.Annotation)` it's actually working fine, great!

File: analyzers/src/SonarAnalyzer.CSharp/Rules/BooleanLiteralUnnecessaryCodeFix.cs [hunk 3]
```diff
@@ -250,73 +251,63 @@ private static Document RemoveConditional(Document document, SyntaxNode root, Co
...
-               || expression is ConditionalExpressionSyntax)
-            {
-                exp = SyntaxFactory.ParenthesizedExpression(expression);
-            }
+        private static ExpressionSyntax GetNegatedExpression(ExpressionSyntax expression) =>
+            SyntaxFactory.PrefixUnaryExpression(SyntaxKind.LogicalNotExpression, AddParenthesisIfNeeded(expression));
 
-            return SyntaxFactory.PrefixUnaryExpression(SyntaxKind.LogicalNotExpression, exp);
-        }
+        private static ExpressionSyntax AddParenthesisIfNeeded(ExpressionSyntax expression) =>
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/BooleanLiteralUnnecessaryCodeFix.cs:310]:
```suggestion
        private static ExpressionSyntax AddParenthesis(ExpressionSyntax expression) =>
```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
The method name is missleading

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**